### DISCUSSION TITLE
What kind of annotations does Artifacthub require for multi-arch container images?

### DISCUSSION THREAD

**Discussion by dploeger** [artifacthub/hub#3761] [Q&A] [answered]
Hi!

I'm currently confused by the annotations, Artifacthub requires for container images as there are index manifest and image manifest annotations.

Apparently, GitHub requires index manifest annotations for multi-arch images (which we use) so that e.g. the description is shown in the package view.

Looking at ArtifactHub's code it seems that it only uses the available archs from the indexmanifest and the metadata from the image manifest.

Where should I put the required annotations?

Thanks!

Kind regards
Dennis

**Comment by tegioz** [comment, marked as answer]
Hi @dploeger 👋

You are right, at the moment AH only uses the annotations in the image manifest (of the default platform) and not the ones in the index manifest. This is something that we'd like to address, relying on the index manifest for multi-arch images. In the meantime, adding them to the image manifest of the `linux/amd64` platform should do the trick.

Hope this helps 🙂

**Reply by dploeger** [reply]
Thanks a lot. I could tackle this in a PR, if you like.

**Reply by dploeger** [reply]
btw: This means adding annotations like `docker build --annotation manifest:io.artifacthub.package.readme-url=https://github.com/dodevops/cloudcontrol`, right?

I'd then add those annotations to each image (as I'm building the multi-arch-image in one go)

**Reply by tegioz** [reply]
No worries!

Please note that [depending on the manifest media type, you may have to use annotations or labels](https://artifacthub.io/docs/topics/repositories/container-images/#image-metadata). 

If you take a look at some of the [container images packages available on AH](https://artifacthub.io/packages/search?kind=12), you can find some inspiration by checking how they do it 🙂 ([sample 1](https://github.com/nginxinc/kubernetes-ingress/blob/432826f81e7ba35074561c73d5d573057d58098a/.github/workflows/build-oss.yml#L180-L184), [sample 2](https://github.com/ublue-os/hwe/blob/b3635f5adb32c0568352c1d5c662346e9ccf9175/.github/workflows/reusable-build.yml#L210-L211)).

A PR for this would be welcome ❤️

**Reply by dploeger** [reply]
Hmm.. I already added the labels in the Dockerfile a while back and AH didn't recognize them apparently. That's why I was switching to annotations now. [This is how](https://github.com/dodevops/cloudcontrol/blob/develop/build/Dockerfile.suffix.mo) the labels were applied (the actual Dockerfile is merged together during [image build](https://github.com/dodevops/cloudcontrol/blob/feature/dpr/annotations/.github/workflows/image.yml#L66))

**Reply by tegioz** [reply]
Labels should work fine, many of the container images listed on `artifacthub.io` use them actually (including the [ah cli tool container image](https://artifacthub.io/packages/container/ah/ah) we publish ourselves ([built using goreleaser, with labels](https://github.com/artifacthub/hub/blob/31cb29e1188d24ada720756b02c8e9ad6a2067f4/.goreleaser.yml#L51-L62))).

BTW please note that readme-url must point to the readme file, not to the root of the repo. Do you have any publicly available image with the required labels that I can take a look at?

**Reply by dploeger** [reply]
The current develop version should have those labels: ghcr.io/dodevops/cloudcontrol-azure:develop

Thanks!

**Reply by tegioz** [reply]
```
% docker image inspect ghcr.io/dodevops/cloudcontrol-azure:develop | jq '.[0] | .Os, .Architecture'
"linux"
"arm64" <--- 😇
```

Please note that AH at the moment looks for those labels in the `linux/amd64` platform.

**Reply by dploeger** [reply]
They're in that one as well:

``` 
DOCKER_DEFAULT_PLATFORM=linux/amd64 docker image inspect ghcr.io/dodevops/cloudcontrol-azure:develop | jq '.[0] | .Os, .Architecture, .Config.Labels'
"linux"
"amd64"
{
  "io.artifacthub.package.readme-url": "https://github.com/dodevops/cloudcontrol",
  "maintainer": "Microsoft",
  "org.label-schema.build-date": "2024-02-29T02:05:49Z",
  "org.label-schema.description": "A great cloud needs great tools; we're excited to introduce Azure CLI, our next generation multi-platform command line experience for Azure.",
  "org.label-schema.docker.cmd": "docker run -v ${HOME}/.azure:/root/.azure -it mcr.microsoft.com/azure-cli:2.58.0",
  "org.label-schema.license": "MIT",
  "org.label-schema.name": "Azure CLI",
  "org.label-schema.schema-version": "1.0",
  "org.label-schema.url": "https://docs.microsoft.com/cli/azure/overview",
  "org.label-schema.usage": "https://docs.microsoft.com/cli/azure/install-az-cli2#docker",
  "org.label-schema.vcs-url": "https://github.com/Azure/azure-cli.git",
  "org.label-schema.vendor": "Microsoft",
  "org.label-schema.version": "2.58.0",
  "org.opencontainers.image.created": "2024-03-12T09:11:57.845Z",
  "org.opencontainers.image.description": "The cloud engineer's toolbox ☁️ 🧰",
  "org.opencontainers.image.documentation": "https://github.com/dodevops/cloudcontrol",
  "org.opencontainers.image.licenses": "MIT",
  "org.opencontainers.image.revision": "1b8ec9dfb61cce73bc9a260a9a1967e04ffd062d",
  "org.opencontainers.image.source": "https://github.com/dodevops/cloudcontrol",
  "org.opencontainers.image.title": "cloudcontrol",
  "org.opencontainers.image.url": "https://github.com/dodevops/cloudcontrol",
  "org.opencontainers.image.vendor": "DO! DevOps",
  "org.opencontainers.image.version": "edge"
}
```

**Reply by tegioz** [reply]
Sorry, my bad 👍

**Reply by dploeger** [reply]
But why doesn't it work then?

**Reply by tegioz** [reply]
I think it's because you're using the OCI manifest format.

```
% crane manifest ghcr.io/dodevops/cloudcontrol-azure:develop | jq .mediaType
"application/vnd.oci.image.index.v1+json"
```

As per the [docs](https://artifacthub.io/docs/topics/repositories/container-images/#image-metadata) I linked before:

> Depending on the image manifest format, metadata must be provided one way or another: images using OCI manifests must use [annotations](https://github.com/opencontainers/image-spec/blob/main/annotations.md), whereas images using Docker V2 manifests must use [config labels](https://docs.docker.com/engine/reference/builder/#label).

**Reply by dploeger** [reply]
Ah okay. I guess, the Github Action switches to that for whatever reason. Then I'm on the right path. Thanks. I'll just add those annotations anywhere I can find. 😆

**Reply by tegioz** [reply]
Hi @dploeger 👋

AH now supports annotations in the index manifest 🙂

Please note that they take precedence over the annotations/labels in the image, just in case you've them in both and they're different.

From <https://artifacthub.io/docs/topics/repositories/container-images/>:

<img width="940" alt="metadata" src="https://github.com/artifacthub/hub/assets/1213902/bd46d9af-5826-4534-9f9c-5b6d40ff021a">

**Reply by dploeger** [reply]
Great. Thanks! ♥️
